### DISCUSSION TITLE
Update-2022-10-11

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#127] [Announcements]
With some trepidation I have merged my text engine branch with the main branch. It is still very much a work in progress, and some menus don't work yet. Please grab the branch and do some tests if you have a chance. There is some minimal documentation in `demo-ui.lisp`.

**Comment by kaveh808** [comment]
Merged some more code.

Renamed some UI files.

Transform menus working. See Demo 05 in `demo-ui.lisp`

![Screen Shot 2022-10-11 at 3 37 54 PM](https://user-images.githubusercontent.com/103860793/195229314-7fd78e92-ea38-4063-949b-3337f86ba80a.png)

**Reply by aykaramba** [reply]
This looks awesome.

**Comment by foretspaisibles** [comment]
Generally speaking, long living branches are a hassle and there is a few methods to avoid them. One of them is feature toggles, maybe we can use them. I can give more details if needed :-)

BTW The screenshot looks very engaging, congrats!

**Reply by kaveh808** [reply]
Yes, I kept adding stuff to the text-engine branch. Should probably have merged it sooner.

What is feature toggles?

**Reply by seletz** [reply]
> Yes, I kept adding stuff to the text-engine branch. Should probably have merged it sooner.
> 
> What is feature toggles?

Not the original author, but generally what I believe Is known as "feature toggles" is that you have some sort of boolean flags where you can turn features on or off.

The intent is that more experimental features can be gated behind such a feature flag which defaults to "off", allowing these changes to evolve in the "main" line of a repository.

HTH.

**Reply by foretspaisibles** [reply]
Yes a feature toggle or feature flag that's exactly what @seletz described, they are a tool avoid long lived branches and the corresponding painful merges. Generally it features a boolean condition of some sort but there is a few tricks and techniques to make them really useful but the general idea is that a long lived branch is always a pain to merge and we therefore need tools to avoid them.

Assume a function

~~~
(define render-scene …
~~~

which is working nicely but someday, you have a new idea to make everything faster, reduce memory consumption, and you really want to try that idea. It will require you to touch a lot of functions and so, maybe making a branch could work but collaborators will also work on the many touch-points your new idea has with all the functions concurring the actual logic of current RENDER-SCENE, and merging could be a difficulty, and maybe turn out to more complex than rewriting from scratch.

To avoid this scenario, it is possible to avoid the long branching.   A basic idea would be to push a new feature in *FEATURES* like `(push :REVOLUTIONARY-RENDER-SCENE *features*)` and put the old code behind `#-REVOLUTIONARY-RENDER-SCENE` and the new code behind `#+REVOLUTIONARY-RENDER-SCENE` but still work on the main branch. Doing so alone is not terribly new — it is really equivalent to just fetch&rebase often — but adding something about the touch-points in the method could help other developers not to break the code you are working on.  For instance, it is possible to write unit tests for the functions in the touch-points and asking people not to break these tests without speaking with you, or it is possible to move all the touching points to the methods for a class, etc.  There is many little different possibilities here but what works is a mix of experience, project convention and advertising correctly the code.

**Reply by kaveh808** [reply]
Thank you for the explanation, @seletz and @foretspaisibles. Good to know.

I hope to avoid long branches in the future. I was lucky in this case as (a) `main` wasn't being updated by anyone else, and (b) I was only touching a couple of text and UI files.

And welcome to `kons-9`, @seletz :)

**Comment by lukego** [comment]
Exciting update!

On my Linux machine it builds and runs but I don't see text on the menus:

![Screenshot from 2022-10-12 11-20-11](https://user-images.githubusercontent.com/13791/195303912-65eab6ce-1495-4c0a-b7b2-f543a12cd969.png)

**Reply by kaveh808** [reply]
Looks like your monitor scale is wrong. Try the steps in issue #125 by setting the monitor scale to 1.
